File: test/scripts/09-random-io/correct.py
#!/usr/bin/env python3
"""
CORRECT: Sequential I/O - enables prefetch, efficient.
Resume: "Prefer Sequential I/O"
"""
import os
import signal
import tempfile
import time
SIZE_MB = 20
CHUNK = 64 * 1024  # 64KB
DURATION_SEC = int(os.environ.get("DURATION_SEC", "300"))
stop = False

def main():
    global stop
    print("CORRECT: Sequential reads")
    signal.signal(signal.SIGINT, lambda s, f: globals().__setitem__('stop', True))
    signal.signal(signal.SIGTERM, lambda s, f: globals().__setitem__('stop', True))
    with tempfile.NamedTemporaryFile(delete=False, suffix='.dat') as f:
        path = f.name
    try:
        with open(path, 'wb') as f:
            f.write(os.urandom(SIZE_MB * 1024 * 1024))
        start = time.perf_counter()
        laps = 0
        while not stop and (time.perf_counter() - start) < DURATION_SEC:
            with open(path, 'rb') as f:
                while f.read(CHUNK):
                    pass
            laps += 1
        print(f"Stopped: {laps} laps in {time.perf_counter()-start:.1f}s")
    finally:
        os.unlink(path)

if __name__ == "__main__":
    main()

File: test/scripts/09-random-io/correct.js
const fs = require('fs');
const path = require('path');
const os = require('os');
const SIZE_MB = 20;
const CHUNK = 64 * 1024;
const DURATION_SEC = parseInt(process.env.DURATION_SEC || '300', 10);
const tmp = path.join(os.tmpdir(), 'perf-seq-io-' + Date.now() + '.dat');
const buf = Buffer.alloc(SIZE_MB * 1024 * 1024);
buf.fill(0);
fs.writeFileSync(tmp, buf);
let laps = 0;
const start = Date.now();
const b = Buffer.alloc(CHUNK);
console.log('CORRECT: Sequential reads');
function run() {
  if (Date.now() - start >= DURATION_SEC * 1000) {
    fs.unlinkSync(tmp);
    console.log('Stopped:', laps, 'laps');
    return;
  }
  const fd = fs.openSync(tmp, 'r');
  while (fs.readSync(fd, b, 0, CHUNK, null) > 0) {}
  fs.closeSync(fd);
  laps++;
  setImmediate(run);
}
run();
process.on('SIGINT', () => { try { fs.unlinkSync(tmp); } catch(e){} process.exit(0); });
